pkg/cloud/scope/shared.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+)
+
+var (
+	// ErrAZSubnetsNotFound is an error when a availability zone is specified but there are
+	// no matching subnets for that availability zone (a.k.a. fault domain)
+	ErrAZSubnetsNotFound = errors.New("no subnets found for supplied availability zone")
+	// ErrLoggerRequired is an error if a logger isn't specified
+	ErrLoggerRequired = errors.New("logger is required")
+	// ErrNotPlaced is an error if there is no placement determined
+	ErrNotPlaced = errors.New("placement not determined")
+)
+
+type placementInput struct {
+	SpecSubnetIDs           []string
+	SpecAvailabilityZones   []string
+	ParentAvailabilityZones []string
+	ControlplaneSubnets     infrav1.Subnets
+}
+
+type subnetsPlacementStratgey interface {
+	Place(input *placementInput) ([]string, error)
+}
+
+func newDefaultSubnetPlacementStrategy(logger logr.Logger) (subnetsPlacementStratgey, error) {
+	if logger == nil {
+		return nil, ErrLoggerRequired
+	}
+
+	return &defaultSubnetPlacementStrategy{
+		logger: logger,
+	}, nil
+}
+
+// defaultSubnetPlacementStrategy is the default strategy for subnet placement
+type defaultSubnetPlacementStrategy struct {
+	logger logr.Logger
+}
+
+// Place works out the subnet placement based on the following precedence:
+// 1. Explicit definition of subnet IDs in the spec
+// 2. If the spec has Availability Zones then get the subnets for these AZs
+// 3. If the parent resource has Availability Zones then get the subnets for these AZs
+// 4. All the subnets from the control plane are used
```

Comment:
Should this be all the subnets or only private ones?
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2191

Reply:
I kept the 'use all control plane subnets' existing behaviour. However, happy to change it to just the private subnets. Is that what we want the default behaviour to be?

Reply:
We default to private subnet in Machines: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/48f64dc6d364d7448bb43c5c7a80bfbc979024a2/pkg/cloud/services/ec2/instances.go#L253

But this may be considered as breaking change.

Reply:
I have made this change to use private subnets. In addition, I have created an `action required` release note so that we can inform users of the behaviour change as part of the release notes.